tests/unit/test_app.py

```diff
@@ -1,30 +1,182 @@
 import os
+from datetime import datetime
+
+import pytest
 
 from email_totals import app
 
+# fixtures for datetime processing around year boundaries
+
+# in december we target nov of this year
+# and compare with oct of this year
+test_now_dec = '2020-12-02'
+
+expected_target_dec = {
+    'Start': '2020-11-01',
+    'End': '2020-12-01',
+}
+
+expected_compare_dec = {
+    'Start': '2020-10-01',
+    'End': '2020-11-01',
+}
+
+# in january we target dec of last year
+# and compare with nov of last year
+test_now_jan = '2020-01-02'
+
+expected_target_jan = {
+    'Start': '2019-12-01',
+    'End': '2020-01-01',
+}
+
+expected_compare_jan = {
+    'Start': '2019-11-01',
+    'End': '2019-12-01',
+}
+
+# in february we target jan of this year
+# and compare with dec of last year
+test_now_feb = '2020-02-02'
+
+expected_target_feb = {
+    'Start': '2020-01-01',
+    'End': '2020-02-01',
+}
+
+expected_compare_feb = {
+    'Start': '2019-12-01',
+    'End': '2020-01-01',
+}
+
+# minimum needs to be a float and will be respected
+minimum = 1.0
+
+
+# https://engineeringfordatascience.com/posts/pytest_fixtures_with_parameterize/
+
+
+@pytest.mark.parametrize(
+    "test_now,expected_target_period,expected_compare_period",
+    [
+        (test_now_dec, expected_target_dec, expected_compare_dec),
+        (test_now_jan, expected_target_jan, expected_compare_jan),
+        (test_now_feb, expected_target_feb, expected_compare_feb),
+    ]
+)
+def test_report_periods(test_now,
+                        expected_target_period,
+                        expected_compare_period):
+    test_dt = datetime.fromisoformat(test_now)
+    found_target, found_compare = app.report_periods(test_dt)
+    assert found_target == expected_target_period
+    assert found_compare == expected_compare_period
+
+
+def test_resource_totals(mocker,
+                         mock_ce_period,
+                         mock_ce_email_target_data,
+                         mock_ce_email_compare_data,
+                         mock_app_resource_dict):
+    mocker.patch('email_totals.ce.get_ce_email_costs',
+                 side_effect=[
+                     mock_ce_email_compare_data,
+                     mock_ce_email_target_data,
+                 ])
+
+    # target and compare periods are passed through to patched functions
+    found_dict = app.get_resource_totals(mock_ce_period,
+                                         mock_ce_period,
+                                         minimum)
+
+    assert found_dict == mock_app_resource_dict
+
+
+def test_account_totals(mocker,
```

Comment:
`get_account_totals` function seems very complicated for just a single test or am I missing something?

Reply:
I did kinda pack a lot into a small test here, the bulk of the logic is actually in `conftest.py`. I tried to define each possible case in those fixtures, for example if you look at `mock_app_account_dict` you can see the response only has entries for 2 of the 4 accounts despite `mock_ce_target_data` having entries for 4 accounts because `mock_org_account_owners` only lists owners for 2 of them.

Reply:
ok